kiwi_lint/similar_string.py

```diff
@@ -0,0 +1,190 @@
+import os
+import re
+import string
+
+import astroid
+from pylint.checkers import BaseChecker, utils
+from pylint.interfaces import IAstroidChecker
+from textdistance import levenshtein
+
+
+class SimilarStringChecker(BaseChecker):
+    __implements__ = (IAstroidChecker,)
+    name = "similar-string-checker"
+
+    msgs = {
+        "R5011": (
+            '"%s" is similar to "%s". Try keeping them the same to reduce work for translators!',
+            "similar-string",
+            "Similar strings should be avoided for single translation",
+        )
+    }
+
+    _dict_of_strings = {}
+    threshold = 0.8
+    error_messages = {}
+
+    project_root = os.path.abspath(
+        os.path.join(os.path.dirname(__file__), "..", "tcms")
+    )
+
+    # NOTE: this works against tcms/ directory and will not take into account
+    # if we want to examine only a sub-dir or a few files
+    # all files found by os.walk
+    all_template_files = set()
+
+    def open(self):
+        for rootpath, _dirs, files in os.walk(self.project_root, topdown=False):
+
+            for file_name in files:
+                if file_name.endswith(".html"):
+                    self.all_template_files.add(
+                        os.path.join(self.project_root, rootpath, file_name)
+                    )
+        for rootpath, _dirs, files in os.walk(
+            os.path.join(self.project_root, "templates", "email"), topdown=False
+        ):
+
+            for file_name in files:
+                if file_name.endswith(".txt"):
+                    self.all_template_files.add(
+                        os.path.join(self.project_root, rootpath, file_name)
+                    )
+
+    @staticmethod
+    def clean_string(text):
+        """
+        This method removes the operators and other punctuations
+        used in the string to avoid unwanted data to add up to
+        the similarity between the strings.
+        """
+        cleaned = ""
+        for char in text:
+            if char not in string.punctuation:
+                cleaned += char
+        return cleaned
+
+    def check_similar_string(self, translation_string):
+        cleaned_translation_string = self.clean_string(translation_string)
+        for key in self._dict_of_strings:
+            if (
+                levenshtein.normalized_similarity(
+                    cleaned_translation_string, self.clean_string(key)
+                )
+                >= self.threshold
+            ):
+                return key
+        return None
+
+    @utils.check_messages("similar-string")
+    def visit_call(self, node):
+        if not (
+            isinstance(node.func, astroid.Name)
+            and node.func.name in ("_", "gettext_lazy")
+        ):
+            return
+        if not isinstance(node.args[0], astroid.nodes.Const):
+            return
+
+        translation_string = node.args[0].value
+
+        if translation_string in self._dict_of_strings:
```

Comment:
From here till the end of this method we're doing essentially the same as `check_similar_and_add_error_message` with minor differences in how the error message is annotated.

For the sake of brevity and reusing the same code you can adapt `check_similar_and_add_error_message` and/or `self.error_messages` to work both with a node/filename (or construct the node from filename inside this method) and make the `close()` method simply iterate over `self.error_messages`. Because you have control over the data structure you can even format it like 
```
{
'node': ...,
'line': ...,
'col_offset': ...
'args': (.....)
}
```

and then `self.add_message("similar-string", **error_message)`.